kotlinx-coroutines-core/common/src/CancellableContinuation.kt

```diff
@@ -198,6 +198,8 @@ public interface CancellableContinuation<in T> : Continuation<T> {
      */
     @ExperimentalCoroutinesApi // since 1.2.0
     public fun resume(value: T, onCancellation: ((cause: Throwable) -> Unit)?)
+
+    public fun <R: T> resume(value: R, onCancellation: ((cause: Throwable, value: R) -> Unit)?)
```

Comment:
It's not a trick: we're notifying the type system that the type of `onCancellation`'s value is decided not by the view of the continuation, but primarily by the type of `value` passed to the function. This type bound promises that `onCancellation` will receive exactly the `value` that this same function receives, and not just any value that `CancellableContinuation` could be resumed with *in theory*.

Without this information, `onCancellation` would have to be ready to accept an arbitrary *supertype* of `T` (effectively, *any value*): because of the `in` variance of `T` in `CancellableContinuation`, `resume` could be called on a more specialized view of `CancellableContinuation`, that is, if we have `CancellableContinuation<Any>`, the call could be `CancellableContinuation<Int>.resume`, and also, in parallel, the continuation could be actually resumed with a `Double`. We know this can't happen, but with the bound, the type system knows that, too.

Reply:
Thanks for the explanation, that's what I've meant :) "Trick" is a substitute for a "a nice and smart solution to our specific issue"